### ISSUE TITLE
Adapt to CNN2

### ISSUE THREAD

**Issue by JakobKHAndersen** [divamgupta/ladder_network_keras#6]
Whoops, accidentally closed the issue.

Basically, I'm trying to replace the fully connected (Dense) layers in the encoder with Conv2D layers in order to classify (28,28,1) input volumes as opposed to (784,) inputs. Hope it makes sense.

**Comment by daniadag** [comment]
did you try it? could you please tell me

**Comment by feiwuu638** [comment]
I tried to use the Conv2D layers as well. Though the model can be trained, there is warning messages "tensorflow:Gradients do not exist for variables……when minimizing the loss", and the testing result is very bad.

**Comment by divamgupta** [comment]
Hi, if could share your Conv2D implementation  then I can take a look.

**Comment by feiwuu638** [comment]
@divamgupta 
Hi, thank you for your attention!

The code is as below (only change the get_ladder_betwork_fc() function):

```
def get_ladder_network_conv(layer_sizes=[784, 1000, 500, 250, 250, 250, 10],
                          noise_std=0.3,
                          denoising_cost=[1000.0, 10.0, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10]):
    L = len(layer_sizes) - 1  # number of layers

    inputs_l = Input((layer_sizes[0],))
    inputs_u = Input((layer_sizes[0],))

    fc_enc = [Dense(s, use_bias=False, kernel_initializer='glorot_normal') for s in layer_sizes[1:]]
    fc_dec = [Dense(s, use_bias=False, kernel_initializer='glorot_normal') for s in layer_sizes[:-1]]
    betas = [AddBeta() for l in range(L + 1)]

    def encoder(inputs, noise_std):
        h = add_noise(inputs, noise_std)
        all_z = [None for _ in range(len(layer_sizes) + 1)]
        all_z[0] = h

        for l in range(1, L + 2):
            if l == 1:
                z_pre = Reshape((28, 28, 1))(h)
                z_pre = Conv2D(filters=3, kernel_size=(3, 3))(z_pre)
                z_pre = Flatten()(z_pre)
            else:
                z_pre = fc_enc[l - 2](h)

            z = Lambda(batch_normalization)(z_pre)
            z = add_noise(z, noise_std)

            if l == L + 1:
                h = Activation('softmax')(betas[l - 1](z))
            else:
                h = Activation('relu')(betas[l - 1](z))

            all_z[l] = z

        return h, all_z

    y_c_l, _ = encoder(inputs_l, noise_std)
    y_l, _ = encoder(inputs_l, 0.0)

    y_c_u, corr_z = encoder(inputs_u, noise_std)
    y_u, clean_z = encoder(inputs_u, 0.0)

    # Decoder
    d_cost = []  # to store the denoising cost of all layers
    for l in range(L, -2, -1):
        z, z_c = clean_z[l+1], corr_z[l+1]
        if l == L:
            u = y_c_u
        elif l > 0:
            u = fc_dec[l](z_est)
        elif l == 0:
            u = Dense(2028, use_bias=False, kernel_initializer='glorot_normal')(z_est)
        else:
            u = Reshape((26, 26, 3))(z_est)
            u = Conv2DTranspose(filters=1, kernel_size=(3, 3))(u)
            u = Flatten()(u)
        u = Lambda(batch_normalization)(u)
        z_est = G_Guass()([z_c, u])
        if l > 0:
            d_cost.append((tf.reduce_mean(tf.reduce_sum(tf.square(z_est - z), 1)) / layer_sizes[l]) * denoising_cost[l + 1])
        elif l == 0:
            d_cost.append((tf.reduce_mean(tf.reduce_sum(tf.square(z_est - z), 1)) / 2028) * denoising_cost[l + 1])
        else:
            d_cost.append((tf.reduce_mean(tf.reduce_sum(tf.square(z_est - z), 1)) / 784) * denoising_cost[l + 1])

    u_cost = tf.add_n(d_cost)

    y_c_l = Lambda(lambda x: x[0])([y_c_l, y_l, y_c_u, y_u, u, z_est, z])

    tr_m = Model([inputs_l, inputs_u], y_c_l)
    tr_m.add_loss(u_cost)
    tr_m.compile(keras.optimizers.Adam(lr=0.02), 'categorical_crossentropy', metrics=['accuracy'])

    tr_m.metrics_names.append("den_loss")
    tr_m.metrics.append(u_cost)

    te_m = Model(inputs_l, y_l)
    tr_m.test_model = te_m

    return tr_m
```

My expected encoder structure is Inputs(784)-Reshape(28,28,1)-Conv2D(26,26,3)-Flatten(2028)-Dense(1000)-Dense(500)-Dense(250)-Dense(250)-Dense(250)-Dense(10), and the decoder structure is the opposite.

The running information is:
```
Epoch 1/5
WARNING:tensorflow:Gradients do not exist for variables ['conv2d_9/kernel:0', 'conv2d_9/bias:0'] when minimizing the loss.
WARNING:tensorflow:Gradients do not exist for variables ['conv2d_9/kernel:0', 'conv2d_9/bias:0'] when minimizing the loss.
WARNING:tensorflow:Gradients do not exist for variables ['conv2d_9/kernel:0', 'conv2d_9/bias:0'] when minimizing the loss.
WARNING:tensorflow:Gradients do not exist for variables ['conv2d_9/kernel:0', 'conv2d_9/bias:0'] when minimizing the loss.
1875/1875 [==============================] - 38s 20ms/step - loss: 13.9158 - accuracy: 0.9822
Epoch 2/5
1875/1875 [==============================] - 39s 21ms/step - loss: 11.5319 - accuracy: 0.9948
Epoch 3/5
1875/1875 [==============================] - 39s 21ms/step - loss: 11.0787 - accuracy: 0.9958
Epoch 4/5
1875/1875 [==============================] - 39s 21ms/step - loss: 10.8780 - accuracy: 0.9966
Epoch 5/5
1875/1875 [==============================] - 39s 21ms/step - loss: 10.6526 - accuracy: 0.9971
test accuracy 0.4393
```

As you can see, there are warnings for no gradients, and the testing accuracy is low with a high training accuracy.

**Comment by ilikewwx** [comment]
> @divamgupta Hi, thank you for your attention!
> 
> The code is as below (only change the get_ladder_betwork_fc() function):
> 
> ```
> def get_ladder_network_conv(layer_sizes=[784, 1000, 500, 250, 250, 250, 10],
>                           noise_std=0.3,
>                           denoising_cost=[1000.0, 10.0, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10]):
>     L = len(layer_sizes) - 1  # number of layers
> 
>     inputs_l = Input((layer_sizes[0],))
>     inputs_u = Input((layer_sizes[0],))
> 
>     fc_enc = [Dense(s, use_bias=False, kernel_initializer='glorot_normal') for s in layer_sizes[1:]]
>     fc_dec = [Dense(s, use_bias=False, kernel_initializer='glorot_normal') for s in layer_sizes[:-1]]
>     betas = [AddBeta() for l in range(L + 1)]
> 
>     def encoder(inputs, noise_std):
>         h = add_noise(inputs, noise_std)
>         all_z = [None for _ in range(len(layer_sizes) + 1)]
>         all_z[0] = h
> 
>         for l in range(1, L + 2):
>             if l == 1:
>                 z_pre = Reshape((28, 28, 1))(h)
>                 z_pre = Conv2D(filters=3, kernel_size=(3, 3))(z_pre)
>                 z_pre = Flatten()(z_pre)
>             else:
>                 z_pre = fc_enc[l - 2](h)
> 
>             z = Lambda(batch_normalization)(z_pre)
>             z = add_noise(z, noise_std)
> 
>             if l == L + 1:
>                 h = Activation('softmax')(betas[l - 1](z))
>             else:
>                 h = Activation('relu')(betas[l - 1](z))
> 
>             all_z[l] = z
> 
>         return h, all_z
> 
>     y_c_l, _ = encoder(inputs_l, noise_std)
>     y_l, _ = encoder(inputs_l, 0.0)
> 
>     y_c_u, corr_z = encoder(inputs_u, noise_std)
>     y_u, clean_z = encoder(inputs_u, 0.0)
> 
>     # Decoder
>     d_cost = []  # to store the denoising cost of all layers
>     for l in range(L, -2, -1):
>         z, z_c = clean_z[l+1], corr_z[l+1]
>         if l == L:
>             u = y_c_u
>         elif l > 0:
>             u = fc_dec[l](z_est)
>         elif l == 0:
>             u = Dense(2028, use_bias=False, kernel_initializer='glorot_normal')(z_est)
>         else:
>             u = Reshape((26, 26, 3))(z_est)
>             u = Conv2DTranspose(filters=1, kernel_size=(3, 3))(u)
>             u = Flatten()(u)
>         u = Lambda(batch_normalization)(u)
>         z_est = G_Guass()([z_c, u])
>         if l > 0:
>             d_cost.append((tf.reduce_mean(tf.reduce_sum(tf.square(z_est - z), 1)) / layer_sizes[l]) * denoising_cost[l + 1])
>         elif l == 0:
>             d_cost.append((tf.reduce_mean(tf.reduce_sum(tf.square(z_est - z), 1)) / 2028) * denoising_cost[l + 1])
>         else:
>             d_cost.append((tf.reduce_mean(tf.reduce_sum(tf.square(z_est - z), 1)) / 784) * denoising_cost[l + 1])
> 
>     u_cost = tf.add_n(d_cost)
> 
>     y_c_l = Lambda(lambda x: x[0])([y_c_l, y_l, y_c_u, y_u, u, z_est, z])
> 
>     tr_m = Model([inputs_l, inputs_u], y_c_l)
>     tr_m.add_loss(u_cost)
>     tr_m.compile(keras.optimizers.Adam(lr=0.02), 'categorical_crossentropy', metrics=['accuracy'])
> 
>     tr_m.metrics_names.append("den_loss")
>     tr_m.metrics.append(u_cost)
> 
>     te_m = Model(inputs_l, y_l)
>     tr_m.test_model = te_m
> 
>     return tr_m
> ```
> 
> My expected encoder structure is Inputs(784)-Reshape(28,28,1)-Conv2D(26,26,3)-Flatten(2028)-Dense(1000)-Dense(500)-Dense(250)-Dense(250)-Dense(250)-Dense(10), and the decoder structure is the opposite.
> 
> The running information is:
> 
> ```
> Epoch 1/5
> WARNING:tensorflow:Gradients do not exist for variables ['conv2d_9/kernel:0', 'conv2d_9/bias:0'] when minimizing the loss.
> WARNING:tensorflow:Gradients do not exist for variables ['conv2d_9/kernel:0', 'conv2d_9/bias:0'] when minimizing the loss.
> WARNING:tensorflow:Gradients do not exist for variables ['conv2d_9/kernel:0', 'conv2d_9/bias:0'] when minimizing the loss.
> WARNING:tensorflow:Gradients do not exist for variables ['conv2d_9/kernel:0', 'conv2d_9/bias:0'] when minimizing the loss.
> 1875/1875 [==============================] - 38s 20ms/step - loss: 13.9158 - accuracy: 0.9822
> Epoch 2/5
> 1875/1875 [==============================] - 39s 21ms/step - loss: 11.5319 - accuracy: 0.9948
> Epoch 3/5
> 1875/1875 [==============================] - 39s 21ms/step - loss: 11.0787 - accuracy: 0.9958
> Epoch 4/5
> 1875/1875 [==============================] - 39s 21ms/step - loss: 10.8780 - accuracy: 0.9966
> Epoch 5/5
> 1875/1875 [==============================] - 39s 21ms/step - loss: 10.6526 - accuracy: 0.9971
> test accuracy 0.4393
> ```
> 
> As you can see, there are warnings for no gradients, and the testing accuracy is low with a high training accuracy.

Hello, I also encountered a similar problem, how did you solve it?
